### ISSUE TITLE
Docfx fails to build documentation for seed project

### ISSUE THREAD

**Issue by DevonKS** [dotnet/docfx#925]
I installed a windows 10 on a new virtual machine (no development environment), downloaded the seed project and downloaded the latest release of docfx. 
When I run docfx with the command:
`docfx.exe docfx-seed-master\docfx.json --serve`

I get this exception:

> Warning: Error opening solution C:/Users/devon/Downloads/docfx-seed-master/docfx-seed-master/src/SampleClass1/SampleClass1.sln: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.. Ignored.
> Warning: No metadata is generated for .
> Info: 5 plug-in(s) loaded.
> Info: Post processor ExtractSearchIndex loaded.
> Info: No files are found with glob pattern **/*.yml, excluding <none>, under directory "C:\Users\devon\Downloads\docfx-seed-master\docfx-seed-master\obj\api"
> Info: [Build Document]Max parallelism is 1.
> Info: [Build Document.CreateMarkdownService]Markdown engine is dfm
> Info: Completed building documents in 846.1755 milliseconds.
> Info: Completed executing in 1441.7506 milliseconds.
> Error: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
>    at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
>    at System.Reflection.RuntimeAssembly.get_DefinedTypes()
>    at System.Composition.Hosting.ContainerConfiguration.<WithAssemblies>b__0(Assembly a)
>    at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
>    at System.Composition.TypedParts.TypedPartExportDescriptorProvider..ctor(IEnumerable`1 types, AttributedModelProvider attributeContext)
>    at System.Composition.Hosting.ContainerConfiguration.CreateContainer()
>    at Microsoft.DocAsCode.Dfm.DfmEngineBuilder..ctor(Options options, String baseDir, String templateDir, IReadOnlyList`1 fallbackFolders)
>    at Microsoft.DocAsCode.Dfm.DocfxFlavoredMarked.CreateBuilder(String baseDir, String templateDir, Options options, IReadOnlyList`1 fallbackFolders)
>    at Microsoft.DocAsCode.Build.Engine.DfmServiceProvider.DfmService..ctor(String baseDir, String templateDir, ImmutableDictionary`2 tokens, IMarkdownTokenTreeValidator tokenTreeValidator, IReadOnlyList`1 fallbackFolders)
>    at Microsoft.DocAsCode.Build.Engine.DfmServiceProvider.CreateMarkdownService(MarkdownServiceParameters parameters)
>    at Microsoft.DocAsCode.Build.Engine.SingleDocumentBuilder.CreateMarkdownService(DocumentBuildParameters parameters, ImmutableDictionary`2 tokens)
>    at Microsoft.DocAsCode.Build.Engine.SingleDocumentBuilder.BuildCore(DocumentBuildParameters parameters)
>    at Microsoft.DocAsCode.Build.Engine.DocumentBuilder.BuildCore(DocumentBuildParameters parameter)
>    at Microsoft.DocAsCode.Build.Engine.DocumentBuilder.Build(IEnumerable`1 parameters, String outputDirectory)
>    at Microsoft.DocAsCode.SubCommands.DocumentBuilderWrapper.BuildDocument(BuildJsonConfig config, TemplateManager templateManager, String baseDirectory, String outputDirectory, String pluginDirectory, String templateDirectory)
>    at Microsoft.DocAsCode.SubCommands.BuildCommand.BuildDocument(String baseDirectory, String outputDirectory)
>    at Microsoft.DocAsCode.SubCommands.BuildCommand.Exec(SubCommandRunningContext context)
>    at Microsoft.DocAsCode.SubCommands.CompositeCommand.Exec(SubCommandRunningContext context)
>    at Microsoft.DocAsCode.Program.ExecSubCommand(String[] args)
> 
> 
> Build failed.
> Warning: Error opening solution C:/Users/devon/Downloads/docfx-seed-master/docfx-seed-master/src/SampleClass1/SampleClass1.sln: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.. Ignored.
> Warning: No metadata is generated for .
> Error: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
>    at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
>    at System.Reflection.RuntimeAssembly.get_DefinedTypes()
>    at System.Composition.Hosting.ContainerConfiguration.<WithAssemblies>b__0(Assembly a)
>    at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
>    at System.Composition.TypedParts.TypedPartExportDescriptorProvider..ctor(IEnumerable`1 types, AttributedModelProvider attributeContext)
>    at System.Composition.Hosting.ContainerConfiguration.CreateContainer()
>    at Microsoft.DocAsCode.Dfm.DfmEngineBuilder..ctor(Options options, String baseDir, String templateDir, IReadOnlyList`1 fallbackFolders)
>    at Microsoft.DocAsCode.Dfm.DocfxFlavoredMarked.CreateBuilder(String baseDir, String templateDir, Options options, IReadOnlyList`1 fallbackFolders)
>    at Microsoft.DocAsCode.Build.Engine.DfmServiceProvider.DfmService..ctor(String baseDir, String templateDir, ImmutableDictionary`2 tokens, IMarkdownTokenTreeValidator tokenTreeValidator, IReadOnlyList`1 fallbackFolders)
>    at Microsoft.DocAsCode.Build.Engine.DfmServiceProvider.CreateMarkdownService(MarkdownServiceParameters parameters)
>    at Microsoft.DocAsCode.Build.Engine.SingleDocumentBuilder.CreateMarkdownService(DocumentBuildParameters parameters, ImmutableDictionary`2 tokens)
>    at Microsoft.DocAsCode.Build.Engine.SingleDocumentBuilder.BuildCore(DocumentBuildParameters parameters)
>    at Microsoft.DocAsCode.Build.Engine.DocumentBuilder.BuildCore(DocumentBuildParameters parameter)
>    at Microsoft.DocAsCode.Build.Engine.DocumentBuilder.Build(IEnumerable`1 parameters, String outputDirectory)
>    at Microsoft.DocAsCode.SubCommands.DocumentBuilderWrapper.BuildDocument(BuildJsonConfig config, TemplateManager templateManager, String baseDirectory, String outputDirectory, String pluginDirectory, String templateDirectory)
>    at Microsoft.DocAsCode.SubCommands.BuildCommand.BuildDocument(String baseDirectory, String outputDirectory)
>    at Microsoft.DocAsCode.SubCommands.BuildCommand.Exec(SubCommandRunningContext context)
>    at Microsoft.DocAsCode.SubCommands.CompositeCommand.Exec(SubCommandRunningContext context)
>    at Microsoft.DocAsCode.Program.ExecSubCommand(String[] args)
>         2 Warning(s)
>         1 Error(s)
> 

I suspect I am missing a hidden dependency for docfx.
Is there anything else I need to install or do to get it working? 
Can the error message be improved to be more meaningful?
Can the documentation be improved to explain what is required to run docfx successfully?

**Comment by DevonKS** [comment]
After installing Microsoft Build Tools 2015 it works. Is there a dependency to this?

**Comment by qinezh** [comment]
@DevonKS Thanks for your feedback, you guess right, it's a dependency to run DocFX, we'll update Documentation later.

**Comment by knaos** [comment]
@qinezh I had the same problem and thanks to the gitter chat I was able to find a solution for that. Please update the documentation.

**Comment by cameron-elliott** [comment]
I don't know that I should comment, but I helped knaos find the issue, can I suggest a more helpful 'fix' is an error in the code about the missing dependency, as opposed to something the user must find in the docs.

**Comment by qinezh** [comment]
@cameron-elliott You're right, it make sense to give some error message in this case.

**Comment by DeluxZ** [comment]
Because of the dependency with Microsoft Build Tools 2015 generating metadata on the (preview) Hosted VS2017 agent in Visual Studio Online fails. The build logs show this line for every project

``Warning:Error opening project C:/a/1/s/src/projectdir/project.csproj: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.. Ignored.``

We are using some C# 7 features which need the Hosted VS2017. Is it easily to remove the dependency with Microsoft Build Tools 2015 ?

**Comment by tzachshabtay** [comment]
I'm seeing the same exception even though I do have VS2015 installed. I also tried explicitly installing Microsoft Build Tools 2015 but it did not help. I downloaded the docfx source and added a try/catch to print the loader exceptions. It shows that it fails to find 'Microsoft.Build.Utilities.Core, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. I looked in my GAC_MSIL folder and I see that the installed version of `Microsoft.Build.Utilities.Core` is v4.0_14.0.0.0__b03f5f7f11d50a3a, so it looks like a versioning issue but I have no idea how to install version 15? I also tried repairing VS2015 which didn't help.

**Comment by qinezh** [comment]
@tzachshabtay  Currently DocFX need VS2017 instead of VS2015, please see [README.md](https://github.com/dotnet/docfx/blob/dev/README.md)

**Comment by tzachshabtay** [comment]
@qinezh I have both VS 2017 (community) and VS 2015 (community).

**Comment by tzachshabtay** [comment]
I managed to solve this (at least for myself).
I'll list everything I did though I'm not sure everything is actually needed:

First I saw that the `Microsoft.Build.Utilities.Core` dll that comes with the binaries is for 15.1.0.0 and the dll that was searched is 15.0.0.0. Then I saw that `Microsoft.DocAsCode.Dfm` doesn't have the M.B.U.C package itself so I tried to install it, this showed me that there were version conflicts. If I used the latest version it conflicted with the version used by `Microsoft.DocAsCode.Metadata.ManagedReference` and if I switched to the before latest version it conflicted with the version used by `Microsoft.DocAsCode.Glob`. I tried to upgrade the version on `ManagedReference` but for some reason nuget restore fails and rolls back. So I went the other way and rolled back the version on `Glob` which compiled.

I then tried running again and... it didn't work, still got the same exception.

So I went to `DfmEngineBuilder` and listed all assemblies that are returned from the linq query in `GetContainer`. M.B.U.C was there in the 15.1.0.0 version only.
I then iterated all of the assemblies and did `GetReferencedAssemblies` one by one to check if somebody is referencing M.B.U.C. As it turns out, there's a library `Microsoft.NuGet.Build.Tasks` which references version 15.0 of M.B.U.C.
I wanted to see if I can maybe upgrade that assembly to use 15.1, but I didn't find where it was used (you have like a bazzilion projects, I tried looking at them one by one but didn't find it). I tried to subscribe to the AppDomain.AssemblyLoad event but it didn't trigger for the Nuget library.
Finally I tried to remove the Nuget library from the list of assemblies given to `WithAssemblies`, i.e added to the linq query: `where !assembly.GetName().FullName.Contains("Microsoft.NuGet.Build.Tasks")`, and after that everything ran to completion and I can see my API and all is good in the world. I don't know if I'm losing any features for removing that assembly, though?
